File: res/jburkardt/people.sc.fsu.edu/~jburkardt/r_src/trapezoidal/trapezoidal.R
trapezoidal <- function ( f, t0, tstop, y0, n, ... ) 

#*****************************************************************************80
#
## trapezoidal() implements the trapezoidal ODE solver.
#
#  Licensing:
#
#    This code is distributed under the GNU LGPL license.
#
#  Modified:
#
#    28 April 2021
#
#  Author:
#
#    Adapted from "cranknic.R, Hans Borchers, the pracma library.
#    Modifications by John Burkardt
#
#  Input:
#
#    function f ( t, y, ... ): evaluates the right hand side of the ODE.
#
#    real T0: the initial time.
#
#    real TSTOP: the final time.
#
#    real Y0: the initial solution values.
#
#    integer N: the number of steps to take.
#
#    ...: optional additional arguments to be passed to F().
#
#  Output:
#
#    list ( T(1:n+1), Y(1:n+1) ): the computed sequence of 
#    solution estimates.
#
{
  stopifnot ( 
    is.numeric(y0),
    is.numeric(t0),
    length(t0) == 1, 
    is.numeric(tstop),
    length(tstop) == 1 )

  if ( is.vector(y0) )
  {
    y0 <- as.matrix(y0)
  }
  else if ( is.matrix(y0) )
  {
    if ( ncol(y0) != 1 ) 
    {
      stop ( "trapezoidal: Argument 'y0' must be a row or column vector.")
    }
  }

  fun <- match.fun(f)
  f <- function(t, y) fun ( t, y, ... )

  y_length <- length ( y0 )
  y <- y0
  yout <- matrix ( NA, n, y_length )
  yout[1, ] <- c(y0)

  dt <- ( tstop - t0 ) / ( n - 1 )
  t <- 0
  ts <- linspace ( t0, tstop, n )
#
#  internal function used for root finding
#
  cnfun <- function(w)  w - y - 0.5* dt * ( f(t, w) + f(t, y) )
  f_length <- length ( f ( t0, y0 ) )
  if ( f_length != y_length )
  {
    stop ( "trapezoidal: Function f must return a vector the same length as 'y0'.")
  }
#
#  Choose the solver used for root finding.
#
  if ( y_length == 1 )
  {
    solver <- fzero
  }
  else
  {
    solver <- fsolve
  }

  for ( i in 2 : n )
  {
    t <- ts[i]
    w <- solver ( cnfun, y )$x
    yout[i, ] <- w
    y <- w
  }

  if ( y_length == 1 ) 
  {
    yout <- drop ( yout )
  }

  return ( list ( t = ts, y = yout ) )
}

File: res/jburkardt/people.sc.fsu.edu/~jburkardt/octave_src/trapezoidal/trapezoidal.m
function [ t, y ] = trapezoidal ( f, tspan, y0, n )

%*****************************************************************************80
%
%% trapezoidal() uses the trapezoidal method + fsolve() to solve an ODE.
%
%  Licensing:
%
%    This code is distributed under the GNU LGPL license.
%
%  Modified:
%
%    26 April 2021
%
%  Author:
%
%    John Burkardt
%
%  Input:
%
%    function handle f: evaluates the right hand side of the ODE.  
%
%    real tspan(2): the starting and ending times.
%
%    real y0(m): the initial conditions. 
%
%    integer n: the number of steps.
%
%  Output:
%
%    real t(n+1,1), y(n+1,m): the solution estimates.
%
  m = length ( y0 );
  t = zeros ( n + 1, 1 );
  y = zeros ( n + 1, m );

  dt = ( tspan(2) - tspan(1) ) / n;

  t(1,1) = tspan(1);
  y(1,:) = y0(:);

  for i = 1 : n

    to = t(i,1);
    yo = y(i,:);

    tn = to + dt; 
    yn = yo + dt * ( f ( to, yo ) )';

    yn = fsolve ( @(yn)trapezoidal_residual(f,to,yo,tn,yn), yn );

    t(i+1,1) = tn;
    y(i+1,:) = yn;

  end

  return
end
function value = trapezoidal_residual ( f, to, yo, tn, yn )

%*****************************************************************************80
%
%% trapezoidal()_residual evaluates the trapezoidal residual.
%
%  Discussion:
%
%    We are seeking a value YN defined by the implicit equation:
%
%      YN = YO + 0.5 * ( TN - TO ) * ( F ( TO, YO ) + F ( TN, YN ) )
%
%  Licensing:
%
%    This code is distributed under the GNU LGPL license.
%
%  Modified:
%
%    15 October 2020
%
%  Author:
%
%    John Burkardt
%
%  Input:
%
%    function handle f: evaluates the right hand side of the ODE.  
%
%    real to, yo: the old time and solution value.
%
%    real tn, yn: the new time and solution value.
%
%  Output:
%
%    real value: the trapezoidal residual.
%
  value = yn - yo - 0.5 * ( tn - to ) * ( f ( to, yo ) + f ( tn, yn ) )';

  return
end

